### ISSUE TITLE
PropertyTypeCoercion fired incorrectly for multi-dimensional array

### ISSUE THREAD

**Issue by michnovka** [vimeo/psalm#7879]
In my code I have:

```php
$view->vars['attr']['class'] = 'form-control';
```

And it results in this error:

```
ERROR: PropertyTypeCoercion - src/Form/Type/AdminEntityType.php:29:13 - $view->vars expects 'array{attr: array<array-key, mixed>, value: mixed}<string, mixed>',  parent type 'array{attr: mixed, value: mixed}<string, mixed>' provided (see https://psalm.dev/198)
```

The variable is defined like this in parent class:

```php
   /**
     * The variables assigned to this view.
     */
    public $vars = [
        'value' => null,
        'attr' => [],
    ];
```

**Comment by michnovka** [comment]
No I cannot, but changing to

```php
$view->vars['attr'] = ['class' => 'form-control'];
```

Fixes the issue, which makes me believe this has to be a bug in psalm.

The FormView is a Symfony class, I cannot edit its PHPDoc

**Comment by AndrolGenhald** [comment]
I've noticed this before too, but I've never had the time to go debug it. Can you try adding a `/** @psalm-trace $view */;' and perhaps `/** @psalm-trace $tmp */ $tmp = $view->vars;`?

> No I cannot, but changing to ... Fixes the issue, which makes me believe this has to be a bug in psalm.

Not necessarily. I don't remember how far I got last time I looked into this, but if Psalm thinks `$view->vars['attr']` might be an `ArrayObject` (ie it's already incorrect) then `$view->vars['attr']['class'] = 'form-control';` will result in the wrong type for `$view->vars['attr']` while `$view->vars['attr'] = ['class' => 'form-control'];` will correct the type.

**Comment by michnovka** [comment]
```
INFO: Trace - src/Form/Type/AdminEntityType.php:27:9 - $tmp: array{attr: array<array-key, mixed>, value: mixed}<string, mixed> (see https://psalm.dev/224)
        /** @psalm-trace $tmp */
        $tmp = $view->vars;
```

**Comment by michnovka** [comment]
Any idea what we can do to address this? Tons of errors whenever I access `$view->vars` :

```
ERROR: PropertyTypeCoercion - src/Form/Type/DateRangeType.php:28:13 - $view->vars expects 'array{attr: array<array-key, mixed>, value: mixed}<string, mixed>',  parent type 'array{attr: mixed, show_clear_button: true, value: mixed}<string, mixed>' provided (see https://psalm.dev/198)
            $view->vars['show_clear_button'] = true;
```

**Comment by michnovka** [comment]
I reproduced it @AndrolGenhald 

https://psalm.dev/r/93d9653415

**Comment by michnovka** [comment]
The issue seems to be the intersection type as changing

```php
     * @psalm-var array{value: ?T, attr: array<array-key, mixed>}&array<string, mixed>
```
for
```php
     * @psalm-var array{value: ?T, attr: array<array-key, mixed>}
```

resolves the issue

**Comment by AndrolGenhald** [comment]
That's definitely a Psalm bug then, Psalm's always behaved a bit weird when intersecting or unioning certain array types.

**Comment by michnovka** [comment]
Is this even a valid definition?

```php
     * @psalm-var array{value: ?T, attr: array<array-key, mixed>}&array<string, mixed>
```

Is this how you specify an array type with multiple keys, but define just some explicitly in `{}`?

**Comment by AndrolGenhald** [comment]
> Is this even a valid definition?

Yes, it's valid because array shapes default to being unsealed, meaning an array with extra keys set still satisfies the type (ie `["foo" => 1, "bar" => 2]` satisfies `array{foo: int}` despite the extra property).

Unfortunately the way arrays and array-shapes are intersected right now has some edge cases, and it's not really possible to correctly express an array with extra keys constrained to a type that doesn't match the defined keys, but Psalm currently just [ignores](https://psalm.dev/r/e0971d3777) this inconsistency. I would like to eventually do something like TypeScript's index signatures like I mentioned [here](https://github.com/vimeo/psalm/issues/5299#issuecomment-1019579981), but I haven't had much time to work on Psalm lately.

**Comment by michnovka** [comment]
Ok, there seems to be some "type flattening" going on: https://psalm.dev/r/2489730359 when working on multi-dimensional arrays

See the [MixedArrayAssignment](https://psalm.dev/117) error which in my opinion is a bug
